### PR TITLE
Implementation of GaussianMixtureMasksTransform

### LINKED ISSUE #1661: Add GaussianMixtureMask transform
See https://github.com/lightly-ai/lightly/issues/1646 for information

### PR DESCRIPTION
**Pull request by snehilchatterjee**
**Modified Components:**
- `lightly/transforms/__init__.py`

**New Files:**
- `lightly/transforms/gaussian_mixture_masks_transform.py`
- `tests/transforms/test_gaussian_mixture_masks.py`

**Enhancement:**
- Added support for Gaussian Mixture Mask transform

**Fixes Issue:**
- Closes #1661

### CODE REVIEW

**Comment by SauravMaheshkar** [comment]
@snehilchatterjee, could you provide some input-output pairs?

**Comment by snehilchatterjee** [comment]
image

File: lightly/transforms/gaussian_mixture_masks_transform.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+        size = f_transform[0].shape
+
+        mask = torch.ones(size)
+
+        for _ in range(num_gaussians):
+            u0 = torch.randint(0, size[0], (1,))
+            v0 = torch.randint(0, size[1], (1,))
+            center = torch.tensor((u0, v0))
+            sigma = torch.rand(2) * 5 + 10
```
Comment (guarin) [inline lightly/transforms/gaussian_mixture_masks_transform.py:82]:
```suggestion
            sigma = torch.rand(2) * (std[1] - std[0]) + std[0]
```

You can also use https://pytorch.org/docs/stable/distributions.html#uniform instead for sampling from uniform distribution. If you use torch distribution please create it in `__init__`

File: lightly/transforms/gaussian_mixture_masks_transform.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+            g_kernel = self.gaussian_kernel((size[0], size[1]), sigma, center)
+            mask -= g_kernel
+
+        filtered_f_transform = f_transform * mask
+        filtered_image = self.irfft2d_transform(filtered_f_transform).abs()
+        return filtered_image
+
+    def __call__(self, image_tensor: Tensor) -> Tensor:
```
Comment (guarin) [inline lightly/transforms/gaussian_mixture_masks_transform.py:91]:
```suggestion
    def __call__(self, freq_image: Tensor) -> Tensor:
```

File: lightly/transforms/gaussian_mixture_masks_transform.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    The mask is created using random Gaussian kernels, which are applied in
+    the frequency domain via RFFT2D, and then the IRFFT2D is used to return
+    to the spatial domain. The transformation is applied to each RGB channel separately.
+
+    Attributes:
+        num_gaussians: Number of Gaussian kernels to generate in the mixture mask.
+        std_range: Tuple containing the minimum and maximum standard deviation for the Gaussians.
+    """
+
+    def __init__(self, num_gaussians: int = 20, std_range: Tuple[int, int] = (10, 15)):
```
Comment (guarin) [inline lightly/transforms/gaussian_mixture_masks_transform.py:23]: Could also be float
```suggestion
    def __init__(self, num_gaussians: int = 20, std_range: Tuple[float, float] = (10, 15)):
```

File: lightly/transforms/gaussian_mixture_masks_transform.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            Tensor: Image after applying the Gaussian mixture mask.
+        """
+        image_size = image_channel[0].shape
+
+        self.irfft2d_transform = IRFFT2DTransform((image_size[0], image_size[1]))
+        f_transform = self.rfft2d_transform(image_channel)
+
+        size = f_transform[0].shape
+
+        mask = torch.ones(size)
```
Comment (guarin) [inline lightly/transforms/gaussian_mixture_masks_transform.py:76]: This makes sure the tensor is on the same device:

```suggestion
        mask = freq_image.new_ones(freq_image.shape)
```

File: lightly/transforms/gaussian_mixture_masks_transform.py (hunk 1 above)
Comment (guarin) [inline lightly/transforms/gaussian_mixture_masks_transform.py:82]: Please pass `device=freq_image.device` whenever creating a tensor.

File: lightly/transforms/gaussian_mixture_masks_transform.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+        Args:
+            size: Tuple specifying the dimensions of the Gaussian kernel (C, H, W).
+            sigma: Tensor specifying the standard deviation of the Gaussian.
+            center: Tensor specifying the center of the Gaussian kernel.
+
+        Returns:
+            Tensor: A 2D Gaussian kernel.
+        """
+        u, v = torch.meshgrid(torch.arange(0, size[0]), torch.arange(0, size[1]))
```
Comment (guarin) [inline lightly/transforms/gaussian_mixture_masks_transform.py:48]: Please set `device` whenever creating new tensors.

**Review by guarin — CHANGES_REQUESTED** [review]
Thanks a lot for the PR, the augmentations look great! I left some comments to make the interface identical to the other FFT transforms.

File: lightly/transforms/gaussian_mixture_masks_transform.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+            v0 = torch.randint(0, size[1], (1,), device=freq_image.device)
+            center = torch.tensor((u0, v0), device=freq_image.device)
+            sigma = torch.rand(2, device=freq_image.device) * (std[1] - std[0]) + std[0]
+
+            g_kernel = self.gaussian_kernel((size[0], size[1]), sigma, center)
+            mask -= g_kernel
+
+        filtered_freq_image = freq_image * mask
+        filtered_image = self.irfft2d_transform(filtered_freq_image).abs()
+        return filtered_image
```
Comment (guarin) [inline lightly/transforms/gaussian_mixture_masks_transform.py:93]: The input and output should be images in the frequency space. The could should be updated like this:

```suggestion
        (C, U, V) = freq_image.shape
        mask = freq_image.new_ones(freq_image.shape)

        for _ in range(num_gaussians):
            u0 = torch.randint(0, U, (1,), device=freq_image.device)
            v0 = torch.randint(0, V, (1,), device=freq_image.device)
            center = torch.tensor((u0, v0), device=freq_image.device)
            sigma = torch.rand(2, device=freq_image.device) * (std[1] - std[0]) + std[0]

            g_kernel = self.gaussian_kernel((U, V), sigma, center)
            mask *= (1 - g_kernel.unsqueeze(0))

        filtered_freq_image = freq_image * mask
        return filtered_freq_image
```

I also had a look at the paper again and I believe all kernels should be multiplied by each other. Otherwise the final mask is not in [0, 1]. With `mask *= (1 - g_kernel)` we always set the center of the kernel to 0. If we would use subtraction instead of multiplication we could get negative numbers.

File: lightly/transforms/gaussian_mixture_masks_transform.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+    def __call__(self, freq_image: Tensor) -> Tensor:
+        """Applies the Gaussian mixture mask transformation to the input frequency-domain image.
+
+        Args:
+            freq_image: Tensor representing a frequency-domain image of shape (C, H, W//2+1).
+
+        Returns:
+            Tensor: The transformed image after applying the Gaussian mixture mask.
```
Comment (guarin) [inline lightly/transforms/gaussian_mixture_masks_transform.py:102]:
```suggestion
            Image tensor in frequency space after applying the Gaussian mixture mask.
```

File: lightly/transforms/gaussian_mixture_masks_transform.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        """
+        transformed_channel: Tensor = self.apply_gaussian_mixture_mask(
+            freq_image, self.num_gaussians, self.std_range
+        )
+        return transformed_channel
```
Comment (guarin) [inline lightly/transforms/gaussian_mixture_masks_transform.py:107]:
```suggestion
        return self.apply_gaussian_mixture_mask(
            freq_image, self.num_gaussians, self.std_range
        )
```

File: tests/transforms/test_gaussian_mixture_masks.py [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import torch
+
+from lightly.transforms import GaussianMixtureMask
+
+
+def test() -> None:
+    transform = GaussianMixtureMask(20, (10, 15))
+    image = torch.rand(3, 32, 17)
+    output = transform(image)
+    assert output.shape == (3, 32, 32)
```
Comment (guarin) [inline tests/transforms/test_gaussian_mixture_masks.py:10]:
```suggestion
    assert output.shape == image.shape
```

File: lightly/transforms/gaussian_mixture_masks_transform.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
+from typing import Tuple
+
+import torch
+import torch.fft
+from torch import Tensor
+
+from lightly.transforms.irfft2d_transform import IRFFT2DTransform
```
Comment (guarin) [inline lightly/transforms/gaussian_mixture_masks_transform.py:7]:
```suggestion
```

**Review by guarin — CHANGES_REQUESTED** [review]
Thanks! Reviewed again

File: lightly/transforms/gaussian_mixture_masks_transform.py (hunk 6 above)
Comment (snehilchatterjee) [reply lightly/transforms/gaussian_mixture_masks_transform.py:93]: Thank you for such a detailed review. I will make the changes today.

**Comment by snehilchatterjee** [comment]
### After making the changes this is the output:
image

**Review by guarin — APPROVED** [review]
This looks great, thanks so much!

**Comment by snehilchatterjee** [comment]
Glad I could contribute
